### PR TITLE
https://github.com/tensorflow/tensorflow/issues/9481

### PR DESCRIPTION
**Pull request by HectorSVC**
Descrition:
Fix the bug that Tensorflow on Windows running into issues when there's UTF8 encoded characters in the file path.
Solution:
Switch to use WideChar API calls for Windows, like the CreateFileW,FindFirstFileW,FindNextFileW, LoadLibraryExW.
Test:
Install Tensorflow on Windows with path have Chinese characers in it. No issue.
Run command "from tensorflow.contrib.rnn.python.ops.gru_ops import *" No issue.

### CODE REVIEW

File: tensorflow/core/platform/windows/windows_file_system.h [hunk 1]
```diff
@@ -66,6 +66,21 @@ class WindowsFileSystem : public FileSystem {
   string TranslateName(const string& name) const override {
     return name;
   }
+
+  std::wstring Utf8ToWideChar(const string& utf8str) {
```
Comment (mrry) [inline tensorflow/core/platform/windows/windows_file_system.h]: Could this be defined as `static`?

File: tensorflow/core/platform/windows/windows_file_system.h [hunk 2]
```diff
@@ -66,6 +66,21 @@ class WindowsFileSystem : public FileSystem {
...
+      int size_required = MultiByteToWideChar(CP_UTF8, 0, &utf8str[0], (int)utf8str.size(), NULL, 0);
+      std::wstring ws_translated_str(size_required, 0);
+      MultiByteToWideChar(CP_UTF8, 0, &utf8str[0], (int)utf8str.size(), &ws_translated_str[0], size_required);
+      return ws_translated_str;
+  }
+
+  string WideCharToUtf8(const std::wstring &wstr) {
```
Comment (mrry) [inline tensorflow/core/platform/windows/windows_file_system.h]: Could this be `static` as well?

File: tensorflow/python/lib/io/file_io_test.py [hunk 3]
```diff
@@ -451,6 +451,12 @@ def testReadlines(self):
     lines = f.readlines()
     self.assertSequenceEqual(lines, data)
 
+  def testUTF8StringPath(self):
```
Comment (mrry) [inline tensorflow/python/lib/io/file_io_test.py]: I'm not sure if I'm reading [PEP 263](https://www.python.org/dev/peps/pep-0263/) correctly, but I think we'll need `# coding=UTF-8 ` near the top of the file. (IIUC, without such a comment, it'll be treated as ASCII.)

**Review by mrry — CHANGES_REQUESTED** [review]
This is looking great, thanks! I just have a few style-related nits.

File: tensorflow/python/lib/io/file_io_test.py (hunk 3 above)
Comment (HectorSVC) [inline tensorflow/python/lib/io/file_io_test.py]: updated

File: tensorflow/core/platform/windows/windows_file_system.h (hunk 1 above)
Comment (HectorSVC) [inline tensorflow/core/platform/windows/windows_file_system.h]: updated

**Review by HectorSVC — COMMENTED** [review]
update code according to review comments

**Review by mrry — APPROVED** [review]
Looks good to me! I'll run the tests and this should be good to merge.

**Comment by mrry** [comment]
Jenkins, test this please!

**Comment by mrry** [comment]
@tensorflow-jenkins test this please.

**Comment by jhseu** [comment]
@HectorSVC mind checking the test failures?

**Comment by HectorSVC** [comment]
looking

**Comment by HectorSVC** [comment]
@jhseu Could you run the test again? I passed these test locally. No clue how it relate to my changes.

**Comment by jhseu** [comment]
Jenkins, test this please

**Comment by vrv** [comment]
@tensorflow-jenkins test this please

**Comment by vrv** [comment]
@tensorflow-jenkins test this please

**Comment by vrv** [comment]
@tensorflow-jenkins test this please

**Comment by vrv** [comment]
@tensorflow-jenkins test this please

**Comment by HectorSVC** [comment]
found more place need to be changed

File: tensorflow/python/lib/io/file_io_test.py [hunk 4]
```diff
@@ -451,6 +452,12 @@ def testReadlines(self):
...
+    file_path = os.path.join(self._base_dir, "UTF8测试_file")
```
Comment (KrisRoofe) [inline tensorflow/python/lib/io/file_io_test.py]: file_path = os.path.join(self._base_dir, "UTF8_test_file")

Comment (HectorSVC) [inline tensorflow/python/lib/io/file_io_test.py]: It's by purpose, to test the file path with utf-8 encoding, to verify the bug fix.

**Comment by HectorSVC** [comment]
@tensorflow-jenkins test this please

**Comment by mrry** [comment]
@tensorflow-jenkins test this please.

**Comment by HectorSVC** [comment]
@mrry, Is there something wrong with Jenkins?

**Comment by mrry** [comment]
Jenkins seems to be working, but I think the code is broken for non-Windows platforms:

https://github.com/tensorflow/tensorflow/blame/dbb9040ca31e25d81482559130e68ca928b19a6c/tensorflow/core/platform/env.cc#L35

This line seems to be including Windows-specific FS code unconditionally into cross-platform code. You probably need to wrap that `#include` in an `#if defined(PLATFORM_WINDOWS)`/`#endif` pair.

**Comment by mrry** [comment]
@tensorflow-jenkins test this please.

**Review by LF-DevJourney — APPROVED** [review]

**Comment by micklexqg** [comment]
tf.gfile.FastGFile(r'F:\train\Rubus_pedatifolius_Genév\4000.jpg', 'rb').read() failed...

**Comment by amapic** [comment]
This correction has been made in which version of tensorflow ?

**Pull request merged**